frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/print/GraphCellPainter.java

```diff
@@ -0,0 +1,128 @@
+package com.virtuslab.gitmachete.frontend.graph.print;
+
+import java.awt.BasicStroke;
+import java.awt.Color;
+import java.awt.Graphics2D;
+import java.awt.RenderingHints;
+import java.awt.geom.Ellipse2D;
+import java.util.Collection;
+
+import com.intellij.vcs.log.graph.EdgePrintElement;
+import com.intellij.vcs.log.graph.NodePrintElement;
+import com.intellij.vcs.log.graph.PrintElement;
+import com.intellij.vcs.log.paint.ColorGenerator;
+import com.intellij.vcs.log.paint.PaintParameters;
+import lombok.RequiredArgsConstructor;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+
+@RequiredArgsConstructor
+public class GraphCellPainter {
+
+  private final ColorGenerator myColorGenerator;
+
+  @UIEffect
+  protected int getRowHeight() {
+    return PaintParameters.ROW_HEIGHT;
+  }
+
+  @UIEffect
+  private BasicStroke getOrdinaryStroke() {
+    return new BasicStroke(PaintParameters.getLineThickness(getRowHeight()), BasicStroke.CAP_ROUND,
+        BasicStroke.JOIN_BEVEL);
+  }
+
+  @UIEffect
+  private void paintUpLine(Graphics2D g2, Color color, int posInRow) {
```

Comment:
Oh, of course, I will comment it there and here too for you.

So the directions (up, down, right) come from the root, the anchor of the edge being drawn. This is because edges do not always go through the whole graph cells straight from up to down. Sometimes the stop in the middle of the cell, at the graph node. 

For clarity have a look at a root branch node. Bearing in mind our nomenclature, it has a down edge but actually never an up edge.

On the other hand, every last sibling branch node has no down edge but it always does have an up edge.

Commit node always has both up and down edges. 

The most complex is the sibling branch with a child node. Here comes a little inconsistency. To this point, you could come to the conclusions that the anchor is identical to this graph node (despite the cases where edges are drawn without nodes, behind other significant nodes AND remembering that commit nodes have invisible nodes aka bullet points). This statement fails here. The anchor stays on the branch indent level, on the left from the bullet point perspective.

So actually so-called anchor is a NODE, this is a "logical part". The bullet point is a "visible part".